bugsnag-android-core/src/main/java/com/bugsnag/android/Configuration.java

```diff
@@ -0,0 +1,821 @@
+     * Add an "on session" callback, to execute code before every
+     * session captured by Bugsnag.
+     *
+     * You can use this to modify sessions before they are stored by Bugsnag.
+     * You can also return <code>false</code> from any callback to ignore a session.
+     *
+     * For example:
+     *
+     * Bugsnag.onSession(new OnSessionCallback() {
+     * public boolean run(Session session) {
+     * return false; // ignore the session
+     * }
+     * })
+     *
+     * @param onSession a callback to run before a session is captured
+     * @see OnSessionCallback
+     */
+    @Override
+    public void addOnSession(@NonNull OnSessionCallback onSession) {
+        if (onSession != null) {
+            impl.addOnSession(onSession);
+        } else {
+            error("addOnSession");
+        }
+    }
+
+    /**
+     * Removes a previously added "on session" callback
+     * @param onSession the callback to remove
+     */
+    @Override
+    public void removeOnSession(@NonNull OnSessionCallback onSession) {
+        if (onSession != null) {
+            impl.removeOnSession(onSession);
+        } else {
+            error("removeOnSession");
+        }
+    }
+
+    /**
+     * Adds a map of multiple metadata key-value pairs to the specified section.
+     */
+    @Override
+    public void addMetadata(@NonNull String section, @NonNull Map<String, ?> value) {
+        if (section != null && value != null) {
+            impl.addMetadata(section, value);
+        } else {
+            error("addMetadata");
+        }
+    }
+
+    /**
+     * Adds the specified key and value in the specified section. The value can be of
+     * any primitive type or a collection such as a map, set or array.
+     */
+    @Override
+    public void addMetadata(@NonNull String section, @NonNull String key, @Nullable Object value) {
+        if (section != null && key != null) {
+            impl.addMetadata(section, key, value);
+        } else {
+            error("addMetadata");
+        }
+    }
+
+    /**
+     * Removes all the data from the specified section.
+     */
+    @Override
+    public void clearMetadata(@NonNull String section) {
+        if (section != null) {
+            impl.clearMetadata(section);
+        } else {
+            error("clearMetadata");
+        }
+    }
+
+    /**
+     * Removes data with the specified key from the specified section.
+     */
+    @Override
+    public void clearMetadata(@NonNull String section, @NonNull String key) {
+        if (section != null && key != null) {
+            impl.clearMetadata(section, key);
+        } else {
+            error("clearMetadata");
+        }
+    }
+
+    /**
+     * Returns a map of data in the specified section.
+     */
+    @Nullable
+    @Override
+    public Map<String, Object> getMetadata(@NonNull String section) {
+        if (section != null) {
+            return impl.getMetadata(section);
+        } else {
+            error("getMetadata");
```

Comment:
I don't think logging makes sense for an accessor method.

Reply:
We need to indicate that something went wrong. How should we do this?

Reply:
I don't have a problem with logging here - checks are based on how the parameters are annotated rather than what the method is for.  We state that the parameter should not be null, so to me it consistent to log an error and noop if this is broken.

Reply:
I think the main reason for logging was to inform the user that a mutation didn't occur as they expected. I'm not sure this applies to a getter, but I'm not too concerned as it applies the rule more consistently :)